File: src/hearth/ui/first_run_offer.py
"""first_run_offer.py — splice the launch page's First-run offer into it.

The entry condition of the first-run path, as a component rather than inline
markup, for the reason every other launch-page asset was extracted: the page
sits at the file-size line, and a component that draws its own state belongs
beside its own bytes. See ``supervisor/firstrun`` for the two facts it draws
and ``routes/state.py`` for where they ride.
"""

from __future__ import annotations

from pathlib import Path

from hearth.ui import pages

#: The placeholder the host page declares in its <script> block.
PLACEHOLDER = "/*FIRST_RUN_JS*/"

PATH = Path(__file__).parent / "first_run_offer.js"

#: The component as of import.
JS = pages.text(PATH)

#: Refuses a page that lost the placeholder — one that shipped it unreplaced
#: would never offer the walk, and would park Start with no way to say why.
splice = pages.splicer(PLACEHOLDER, PATH, "the first-run offer",
                       "in its <script> block")

File: src/hearth/ui/first_run_offer.js
// first_run_offer.js — the launch page's First-run offer.
//
// The entry condition of the first-run path: while the selected model's id is
// still the shipped placeholder, or nothing on this install has a session yet,
// the launch page points at /admin/first-run — and while the id is the
// placeholder it parks its own Start, because a bot started now would ask the
// server for a model that does not exist. Both facts ride /admin/state as
// first_run: {needs_model, fresh}; an install too broken to say answers null
// and this draws nothing. Host page supplies a #firstrun mount and calls
// render(state); the return value is "keep Start parked".
window.firstRun = (function () {
  function render(data) {
    const fr = (data && data.first_run) || null;
    const on = !!(fr && (fr.needs_model || fr.fresh));
    const host = document.getElementById("firstrun");
    host.classList.toggle("hidden", !on);
    if (!on) return false;
    host.replaceChildren();
    const h = document.createElement("h2");
    h.textContent = "First run";
    const p = document.createElement("p");
    p.textContent = fr.needs_model
      ? "The selected model config still carries the shipped placeholder id, so " +
        "Start is parked. The first-run page lists what your server serves and " +
        "records your pick."
      : "Nothing has been said on this install yet. The first-run page walks the " +
        "first sitting: server, model, companion, and the first words.";
    const a = document.createElement("a");
    a.href = "/admin/first-run";
    a.textContent = "Walk through the first run →";
    const q = document.createElement("p");
    q.appendChild(a);
    host.appendChild(h); host.appendChild(p); host.appendChild(q);
    return !!fr.needs_model;
  }
  return { render: render };
})();
